### PR TITLE
feat: typescript changes (#39520)

### PR DESCRIPTION
**Pull request by Aamil-Successive**
Description

Create a new folder extraTs.
Define tsconfig.json.
Config is:

{
"compilerOptions": {
"module": "commonjs",
"target": "es6",
"noImplicitAny": false,
"moduleResolution": "node",
"allowSyntheticDefaultImports": true,
"resolveJsonModule": true,
"esModuleInterop": false,
"allowJs": true /* Allow javascript files to be compiled. */,
"checkJs": false /* Report errors in .js files. */,
"sourceMap": true,
"outDir": "./dist",
"baseUrl": ".",
"paths": {
"*": [
"node_modules/*",
"src/types/*"
]
},
},
"exclude": ["dist"],
"include": [
"extraTs"
],
"types": ["src/types"]
}

Install dependencies(as dev) for running typescript:
- typescript
- tsc-watch

-nodemon

Copy the folders (patterns and utils) and files (constants and index.js) from extra to extraTs.
Rename all files to .ts extension.
Also, create a file interfaces.ts to put all interfaces.
Now convert all the code to typescript as described in the class.

Add dist folder to .gitignore.

Acceptance Criteria:
- Every function must have types defined for arguments and also the return value.

Commands(Scripts inside package.json): 
 

"start:extra": "babel-node extra",

"start:extraTs": "tsc-watch --project . --onSuccess \"npm run nodemon\"",

"nodemon": "nodemon ./dist/index.js",

"test": "echo \"Error: no test specified\" && exit 1",

"lint": "./node_modules/tslint/bin/tslint -c tslint.json -p tsconfig.json --exclude 'node_modules/**/*.{j,t}s'",

"lint:fix": "npm run lint --fix"

- Also setup tslint, husky, and commitzen
Note: Husky must be implemented with both pre-commit and pre-push hooks, and must run command ```npm run lint```

 

tslint.json file config:

### CODE REVIEW

File: extraTs/index.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { diamond, equilateral } from './patterns';
+
+import { hasPermission, validateUsers } from  './utils';
+
+const users = [
```
Comment (adilsuccessive) [inline extraTs/index.ts]: users type missing.

File: extraTs/patterns/diamond.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+export default function diamond(a) {
```
Comment (adilsuccessive) [inline extraTs/patterns/diamond.ts]: return type of function and argument 'a' type is missing.

File: extraTs/patterns/equilateral.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+export default function equilateral (a) {
```
Comment (adilsuccessive) [inline extraTs/patterns/equilateral.ts]: return type of function and argument 'a' type is missing.

File: extraTs/utils/permissions.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { permissions } from '../constants';
+const hasPermission = (moduleName: string, Role: string, PERMISSION_TYPE: string) => {
```
Comment (adilsuccessive) [inline extraTs/utils/permissions.ts]: take proper arguments name  for .
Role , PERMISSION_TYPE

Comment (adilsuccessive) [reply extraTs/utils/permissions.ts]: return type is also missing for hasPermission

File: extraTs/utils/validation.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { validateEmail } from './helpers';
+
+    const validateUsers = (users) => {
```
Comment (adilsuccessive) [inline extraTs/utils/validation.ts]: users type and return type is missing.

**Review by adilsuccessive — CHANGES_REQUESTED** [review]

File: extraTs/index.ts (hunk 1 above)
Comment (Aamil-Successive) [reply extraTs/index.ts]: Changes Done.

File: extraTs/patterns/diamond.ts (hunk 2 above)
Comment (Aamil-Successive) [reply extraTs/patterns/diamond.ts]: Changes Done.

File: extraTs/patterns/equilateral.ts (hunk 3 above)
Comment (Aamil-Successive) [reply extraTs/patterns/equilateral.ts]: Changes Done.

File: extraTs/utils/permissions.ts (hunk 4 above)
Comment (Aamil-Successive) [reply extraTs/utils/permissions.ts]: Changes Done.

File: extraTs/utils/validation.ts (hunk 5 above)
Comment (Aamil-Successive) [reply extraTs/utils/validation.ts]: Changes Done.

**Review by adilsuccessive — COMMENTED** [review]
Still the return type of all the 4 functions (diamond, equilateral, hasPermission and validateUser) missing.

**Review by adilsuccessive — APPROVED** [review]

**Pull request merged**